src/window_manipulation.rs

```diff
@@ -0,0 +1,374 @@
+	Resizing {
+		/// The window this window manipulation applies to.
+		window: Window,
+		/// The original dimensions of the `window`.
+		orig_size: (u16, u16),
+		/// The starting position of the cursor when the window manipulation begins.
+		///
+		/// The `cursor_pos` is relative to the `window`'s root window.
+		///
+		/// This is used to calculate how far the cursor has moved during window manipulation, and
+		/// thus the offset that should be applied to the `window`'s dimensions.
+		cursor_pos: (i16, i16),
+	},
+}
+
+#[allow(dead_code)]
+impl WindowManipulation {
+	/// Creates a [`WindowManipulation::Moving`] window manipulation.
+	///
+	/// Blocking: waits for one [GetGeometry](xcb::x::GetGeometry) request, one
+	/// [QueryTree](xcb::x::QueryTree) request.
+	///
+	/// Flushes the connection.
+	pub fn moving(conn: &Connection, window: Window, cursor_pos: (i16, i16)) -> xcb::Result<Self> {
+		// Get geometry request to get the window's coordinates.
+		let geometry_req = get_geometry(conn, window);
+
+		// Window tree request to get the window's parent.
+		trace!(window = window.resource_id(), "Requesting window tree");
+		let tree_req = conn.send_request(&x::QueryTree { window });
+
+		conn.flush()?;
+
+		let geometry = conn.wait_for_reply(geometry_req)?;
+		let tree = conn.wait_for_reply(tree_req)?;
+
+		debug!(
+			window = window.resource_id(),
+			"Beginning window manipulation"
+		);
+		Ok(Self::Moving {
+			window,
+			parent: tree.parent(),
+			orig_coords: (geometry.x(), geometry.y()),
+			cursor_pos,
+		})
+	}
+
+	/// Creates a [`WindowManipulation::Resizing`] window manipulation.
+	///
+	/// Blocking: waits for one [GetGeometry](xcb::x::GetGeometry) request.
+	///
+	/// Flushes the connection.
+	pub fn resizing(
+		conn: &Connection,
+		window: Window,
+		cursor_pos: (i16, i16),
+	) -> xcb::Result<Self> {
+		// Get geometry request to get the window's dimensions.
+		let geometry = conn.wait_for_reply(get_geometry(conn, window))?;
+
+		debug!(
+			window = window.resource_id(),
+			"Beginning window manipulation"
+		);
+		Ok(Self::Resizing {
+			window,
+			orig_size: (geometry.width(), geometry.height()),
+			cursor_pos,
+		})
+	}
+
+	/// Returns whether this [WindowManipulation] is [`Moving`](WindowManipulation::Moving).
+	///
+	/// This is the inverse of [WindowManipulation::is_resizing()], and is simply another shorthand
+	/// for:
+	/// ```rust
+	/// matches!(self, Self::Moving { .. })
+	/// ```
+	///
+	/// See also: [`matches!`]
+	pub fn is_moving(self) -> bool {
+		matches!(self, Self::Moving { .. })
+	}
+
+	/// Returns whether this [WindowManipulation] is [`Resizing`](WindowManipulation::Resizing).
+	///
+	/// This is the inverse of [WindowManipulation::is_moving()], and is simply another shorthand
+	/// for:
+	/// ```rust
+	/// matches!(self, Self::Resizing { .. })
+	/// ```
+	///
+	/// See also: [`matches!`]
+	pub fn is_resizing(self) -> bool {
+		matches!(self, Self::Resizing { .. })
+	}
+
```

Comment:
Wouldn't these two functions make more sense as macros if possible? They may add a little unnecessary overhead for simple shorthands

Reply:
Yes, you're correct. I'll make sure to change them to macros. I only did them as functions because I'm not used to macros yet and didn't think of it - my mistake.

Reply:
I don't think they would make more sense as macros, _but_ they would make sense as `const fn`s. As far as I can tell, that's the same kind of idea - `const fn`s are replaced at compile time, similar to macros but they aren't targeting syntax directly. `std`'s `Option` actually has pretty much the exact same thing as `const fn`s:
```rust
pub const fn is_some(&self) -> bool {
    matches!(*self, Some(_))
}
```
So I will change them to `const fn`.